File: src/content/3.5/code/scala/snippet19.scala
implicit def stateMonad[S] = new Monad[State[S, ?]] {
  def flatMap[A, B](sa: State[S, A])(k: A => State[S, B]): State[S, B] =
    State { s =>
      val (a, s1) = runState(sa)(s)
      runState(k(a))(s1)
    }
  
  def pure[A](a: A): State[S, A] = State(s => (a, s))
}

File: src/content/3.5/code/ocaml/snippet19.ml
module State_Monad (S : sig
  type t
end) : Monad_Bind = struct
  type 'a m = (S.t, 'a) state

  let ( >>= ) sa k =
    State
      (fun s ->
        let a, s' = run_state sa s in
        let sb = k a in
        run_state sb s')


  let return a = State (fun s -> a, s)
end
